example/src/main/java/io/netty/example/dns/doh/ALPNHandler.java

```diff
@@ -0,0 +1,96 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.example.dns.doh;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.dns.doh.DoHDecoder;
+import io.netty.handler.codec.dns.doh.DoHEncoder;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpObjectAggregator;
+import io.netty.handler.codec.http2.DefaultHttp2Connection;
+import io.netty.handler.codec.http2.DelegatingDecompressorFrameListener;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.HttpToHttp2ConnectionHandler;
+import io.netty.handler.codec.http2.HttpToHttp2ConnectionHandlerBuilder;
+import io.netty.handler.codec.http2.InboundHttp2ToHttpAdapter;
+import io.netty.handler.codec.http2.InboundHttp2ToHttpAdapterBuilder;
+import io.netty.handler.ssl.ApplicationProtocolNames;
+import io.netty.handler.ssl.ApplicationProtocolNegotiationHandler;
+import io.netty.util.concurrent.Promise;
+import io.netty.util.internal.ObjectUtil;
+
+import java.net.URL;
+
+final class ALPNHandler extends ApplicationProtocolNegotiationHandler {
+
+    private final URL url;
+    private final Promise<Void> promise;
+
+    ALPNHandler(URL url, Promise<Void> promise) {
+        super(ApplicationProtocolNames.HTTP_1_1);
+        this.url = url;
+        this.promise = ObjectUtil.checkNotNull(promise, "promise");
+    }
+
+    Promise<Void> promise() {
+        return promise;
+    }
+
+    @Override
+    protected void configurePipeline(ChannelHandlerContext ctx, String protocol) {
+        if (protocol.equalsIgnoreCase(ApplicationProtocolNames.HTTP_2)) {
+            Http2Connection connection = new DefaultHttp2Connection(false);
+
+            InboundHttp2ToHttpAdapter listener = new InboundHttp2ToHttpAdapterBuilder(connection)
+                    .propagateSettings(false)
+                    .validateHttpHeaders(true)
+                    .maxContentLength(1024 * 64)
+                    .build();
+
+            HttpToHttp2ConnectionHandler http2Handler = new HttpToHttp2ConnectionHandlerBuilder()
+                    .frameListener(new DelegatingDecompressorFrameListener(connection, listener))
+                    .connection(connection)
+                    .build();
+
+            ctx.pipeline().addLast(http2Handler,
+                    new HttpObjectAggregator(1024 * 64, true),
+                    new DoHEncoder(url),
+                    new DoHDecoder(),
+                    new Handler());
+            promise.setSuccess(null);
+        } else if (protocol.equalsIgnoreCase(ApplicationProtocolNames.HTTP_1_1)) {
```

Comment:
HTTP/2 is the minimum recommended protocol here, HTTP/1.1 is not a great fallback

Should this be optional/configurable?

https://tools.ietf.org/html/rfc8484#section-5.2

Reply:
I did think about this but `RECOMMENDED` does not enforce (as per my best knowledge). If it was `MUST` then HTTP/1.1 was not allowed.

Please correct if I'm wrong.

Reply:
@slandelle You have any idea about this `RECOMMENDED` thing? I am also confused now.

Reply:
My reading, is that if you fallback to HTTP/1.1 then likely you have a very active bottleneck in your system and using DNS over HTTPS will effectively be broken for real world cases. Yes it works, but if something like a intermediate proxy is dropping you down from HTTP/2 to HTTP/1.1 then clients may either want to be aware, notified, or make it configurable.

I guess you can restrict in SslContextBuilder.forClient

Reply:
I completely agree with that. HTTP/1.1 will ruin the performance. DoT can shine out in that case but out of context for now. RFC says `RECOMMENDED` along with `MINIMUM` HTTP/2.

I ran queries over HTTP/1.1 and they were successful (both Cloudflare and Google DNS).